mlrun/runtimes/base.py

```diff
@@ -315,102 +315,39 @@ def run(
-            runspec = deepcopy(runspec)
-            if isinstance(runspec, str):
-                runspec = literal_eval(runspec)
-            if not isinstance(runspec, (dict, RunTemplate, RunObject)):
-                raise ValueError(
-                    "task/runspec is not a valid task object," f" type={type(runspec)}"
-                )
-
-        if isinstance(runspec, RunTemplate):
-            runspec = RunObject.from_template(runspec)
-        if isinstance(runspec, dict) or runspec is None:
-            runspec = RunObject.from_dict(runspec)
+        runspec = self._resolve_runspec_runtime(runspec)
 
-        runspec.spec.handler = (
-            handler or runspec.spec.handler or self.spec.default_handler or ""
-        )
-        if runspec.spec.handler and self.kind not in ["handler", "dask"]:
-            runspec.spec.handler = runspec.spec.handler_name
-
-        def_name = self.metadata.name
-        if runspec.spec.handler_name:
-            short_name = runspec.spec.handler_name
-            if "::" in short_name:
-                short_name = short_name.split("::")[1]  # drop class name
-            def_name += "-" + short_name
-        runspec.metadata.name = name or runspec.metadata.name or def_name
-        verify_field_regex(
-            "run.metadata.name", runspec.metadata.name, mlrun.utils.regex.run_name
-        )
-        runspec.metadata.project = (
-            project
-            or runspec.metadata.project
-            or self.metadata.project
-            or config.default_project
-        )
-        runspec.spec.parameters = params or runspec.spec.parameters
-        runspec.spec.inputs = inputs or runspec.spec.inputs
-        runspec.spec.hyperparams = hyperparams or runspec.spec.hyperparams
-        runspec.spec.hyper_param_options = (
-            hyper_param_options or runspec.spec.hyper_param_options
-        )
-        runspec.spec.verbose = verbose or runspec.spec.verbose
-        if scrape_metrics is None:
-            if runspec.spec.scrape_metrics is None:
-                scrape_metrics = config.scrape_metrics
-            else:
-                scrape_metrics = runspec.spec.scrape_metrics
-        runspec.spec.scrape_metrics = scrape_metrics
-        runspec.spec.output_path = out_path or artifact_path or runspec.spec.output_path
-        runspec.spec.input_path = (
-            workdir or runspec.spec.input_path or self.spec.workdir
+        runspec = self._enrich_runspec(
+            runspec,
+            handler,
+            project,
+            name,
+            params,
+            inputs,
+            hyperparams,
+            hyper_param_options,
+            verbose,
+            scrape_metrics,
+            out_path,
+            artifact_path,
+            workdir,
         )
-        if self.spec.allow_empty_resources:
-            runspec.spec.allow_empty_resources = self.spec.allow_empty_resources
-
         spec = runspec.spec
-        if spec.secret_sources:
```

Comment:
In the line above there is `spec = runspec.spec` which (IMO) just makes reading harder cause you're seeing `spec` but don't know a spec of what it is, should simply remove that and use `run.spec` everywhere
Same is true for `meta` that is defined below (let's just use `run.metadata`